src/jit/lowerarmarch.cpp

```diff
@@ -348,6 +343,66 @@ void Lowering::LowerBlockStore(GenTreeBlk* blkNode)
     }
 }
 
+//------------------------------------------------------------------------
+// ContainBlockStoreAddress: Attempt to contain an address used by an unrolled block store.
+//
+// Arguments:
+//    blkNode - the block store node
+//    size - the block size
+//    addr - the address node to try to contain
+//
+void Lowering::ContainBlockStoreAddress(GenTreeBlk* blkNode, unsigned size, GenTree* addr)
+{
+    assert(blkNode->OperIs(GT_STORE_BLK) && (blkNode->gtBlkOpKind == GenTreeBlk::BlkOpKindUnroll));
+    assert(size < INT32_MAX);
+
+    if (addr->OperIsLocalAddr())
+    {
+        addr->SetContained();
+        return;
+    }
+
+    if (!addr->OperIs(GT_ADD) || addr->gtOverflow() || !addr->AsOp()->gtGetOp2()->OperIs(GT_CNS_INT))
```

Comment:
Why `TryCreateAddrMode` can't be used here as it happens on XARCH? 
I see that it does different checks, but it is unclear if this case is unique and other calls to `TryCreateAddrMode` on arm are correct or all other calls to ``TryCreateAddrMode` need to be changed to check `if ((size >= 2 * REGSIZE_BYTES) && (offset % REGSIZE_BYTES != 0))` for example.

Reply:
The address mode situation on ARM is pretty messy - `TryCreateAddrMode` (well, `genCreateAddrMode` really), was written for x86 and despite having tons of ifdefs in it (and I've rarely seen such an abuse of ifdefs) still generates address modes that aren't really valid on ARM and relies on the emitter to sort things out - for example: https://github.com/dotnet/coreclr/blob/5a381e7c8507777f71746ec564a6b75e38ebe089/src/jit/emitarm64.cpp#L11823

And the emitter sorts things out only in certain cases (if you use `emitInsLoadStoreOp` and that only works with `GT_IND` nodes, does not support `LDP`/`STP`, requires temporary registers) so it's not suitable for block init/copy purposes.

So I preferred to avoid `TryCreateAddrMode`. It may be possible to use it to create an address mode node and then do extra checks before containing it. But there doesn't seem to be any point in doing that - it's a lot of extra code that needs to run only to produce a `LEA` that may not be useful and the original `ADD` should have been left alone.